### PR TITLE
Add tests for new String.Join overloads 

### LINKED ISSUE #5552: Add string.Join(char, ...) overloads
Moved from https://github.com/dotnet/coreclr/issues/2736#issuecomment-173295835 on behalf of @jasonwilliams200OK 

Current situation:

``` c#
char separator = ':';
string joined = string.Join(separator.ToString(), "one", "two", "three");
```

With `char` overload, users will be able to join string without calling `ToString()`
i.e. `string joined = string.Join(separator, "one", "two", "three");`

This will reduce at least one string allocation per call: https://github.com/dotnet/coreclr/blob/80c2952/src/mscorlib/src/System/Char.cs#L145-L168.
# 

Proposed API surface - @jasonwilliams200OK 

``` csharp
// in -> coreclr/src/mscorlib/src/System/String.cs
public static string Join(char separator, params object[] value)
public static string Join<T>(char separator, IEnumerable<T> values)
public static string Join(char separator, params string[] value)
public static string Join(char separator, string[] value, int startIndex, int count)

// REJECTED:
//public static string Join(char separator, IEnumerable<string> values)

// in -> coreclr/src/mscorlib/src/System/UnSafeCharBuffer.cs
public void AppendString(char charToAppend)
```

### PR DESCRIPTION
**Pull request by AlexRadch**
Close #5552

### CODE REVIEW

File: src/System.Runtime/tests/System/StringTests.cs [hunk 1]
```diff
@@ -1653,6 +1653,51 @@ public static void Join_ObjectArray_Null_ThrowsArgumentNullException()
...
+            new object[] { '$', new object[] { "", "Bar", "", }, 0, 3, "$Bar$" },
+            new object[] { '$', new object[] { "Foo", "Bar", "Baz" }, 0, 3, "Foo$Bar$Baz" },
+
+            new object[] { '$', new object[] { "Foo", "Bar", "Baz" }, 0, 3, "" },
+            new object[] { '$', new object[] { "Foo", "Bar", "Baz" }, 1, 1, "Bar" },
+        };
+
+        [Theory]
+        [MemberData(nameof(Join_Char_TestData))]
+        public static void Join_Char_Test(char separator, object[] values, int startIndex, int count, string expected)
```
Comment (justinvp) [inline src/System.Runtime/tests/System/StringTests.cs]: Shouldn't these tests go in a new file named `StringTests.netcoreapp1.1.cs` that's conditionally included?

Comment (AlexRadch) [inline src/System.Runtime/tests/System/StringTests.cs]: @justinvp I tried make so. But separated file with conditional compilation was not compiled. I am think how correct include conditional compilation.

**Comment by jamesqo** [comment]
@AlexRadch, can you add tests for argument validation for the new APIs too? e.g. Making sure they throw the correct exceptions with the correct parameter names? I already wrote some tests for that in https://github.com/dotnet/corefx/pull/14631, so you can just copy and paste them from there if desired.

File: src/System.Runtime/tests/System/StringTests.cs [hunk 2]
```diff
@@ -7,14 +7,15 @@
...
+using System.Linq;
 using System.Runtime.InteropServices;
 using System.Threading;
 using System.Threading.Tasks;
 using Xunit;
 
 namespace System.Tests
 {
-    public static class StringTests
+    public static partial class StringTests
```
Comment (joperezr) [inline src/System.Runtime/tests/System/StringTests.cs]: As @justinvp suggests, please do create a new file called StringTests.netcoreapp1.1.cs and move the netcoreapp1.1-specific tests in there. On the .csproj, add this file on the ItemGroup that is conditioned for netcoreapp1.1 only.

**Comment by karelz** [comment]
@AlexRadch do you plan to finish the work?

**Comment by karelz** [comment]
The PR seems to be abandoned for 1 month. Closing for now.

**Pull request closed without merging**